### PR TITLE
Upgrade to SDK v0.43+

### PR DESCRIPTION
**Pull request by ancazamfir**
Closes: #947
Closes: #918

## Description

The proto files in this branch are currently aligned with:

- ibc-go v1.0.0-alpha2: [d70f49c8f612d60f1b7e2f1d1f160f28988962e1](https://github.com/cosmos/ibc-go/commit/d70f49c8f612d60f1b7e2f1d1f160f28988962e1)
- cosmos-sdk v0.43.0-beta1: [8cfa2c26738276d895caf9eb98b3f70616218e17](https://github.com/cosmos/cosmos-sdk/commit/8cfa2c26738276d895caf9eb98b3f70616218e17)

______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [x] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Comment by colin-axner** [comment]
Did some testing using this branch between a gaia bin on v4.2.1 and the hub/cosmoshub-4.6 branch. Everything works great! I do get some "packet already received" errors when running `hermes start-multi`. See [logs](https://gist.github.com/colin-axner/28cc51b6690d99250d358459bde7ea8c). Is this expected?

The main downside is lost fees for retrying these transactions

**Comment by ancazamfir** [comment]
> Did some testing using this branch between a gaia bin on v4.2.1 and the hub/cosmoshub-4.6 branch. Everything works great! I do get some "packet already received" errors when running `hermes start-multi`. See [logs](https://gist.github.com/colin-axner/28cc51b6690d99250d358459bde7ea8c). Is this expected?
> 
> The main downside is lost fees for retrying these transactions
> 
> Edit: It works fine if I use gaiad to initiate the transfer. Must be related to `hermes tx raw`

This is not expected unless you had another relayer running in parallel. Could you give more detail in the setup? And the exact `ft-transfer` command you used? I use this command often and haven't seen this before.

**Comment by colin-axner** [comment]
I setup an old/new gaia using the gaiad manager. Created channels and sent tokens using ts-relayer. Then stopped the ts-relayer (I'm fairly certain)

Created channels using hermes, started hermes via start-multi. Sent using
```
hermes tx raw ft-transfer network2 network1 transfer channel-0 1 -o 1000 -n 1 -d localcoin
```
got the above logs.

Then sent using gaiad afterwards and it worked fine. 

I will try to reproduce again later today

**Comment by colin-axner** [comment]
I managed to reproduce. The problem scenario is when the ts-relayer is also running. Specifically:
- start ts-relayer
- send packet
- wait for ts-relayer to log that it has relayed the packet
- start hermes

Then hermes will actually beat the ts-relayer to the acknowledgement process but get stuck on trying to relay the packet

Does hermes rely on the ibc-go error codes at all? We are planning on making the error numbers as apart of our API for versioning so we don't change them often and when we do it is well documented. We can turn the unordered channel [packet already received](https://github.com/cosmos/ibc-go/blob/f4123347754d814759c9208273fe6772880063cb/modules/core/04-channel/keeper/packet.go#L244) into an error code so that hermes could stop retries. Would this help?

**Comment by ancazamfir** [comment]
> Does hermes rely on the ibc-go error codes at all? We are planning on making the error numbers as apart of our API for versioning so we don't change them often and when we do it is well documented. We can turn the unordered channel [packet already received](https://github.com/cosmos/ibc-go/blob/f4123347754d814759c9208273fe6772880063cb/modules/core/04-channel/keeper/packet.go#L244) into an error code so that hermes could stop retries. Would this help?

Thanks @colin-axner ! We are working on better error filtering and definitely having the error code here will help. 

BTW, this also happens with two hermes relayers running. We haven't really started doing these types of tests heavily and understand the issues. But it's coming.

**Comment by colin-axner** [comment]
opened a pr to update the [error messages](https://github.com/cosmos/ibc-go/pull/184), also opened an issue on the [spec repo](https://github.com/cosmos/ibc/issues/574)

Please checkout the ibc-go pr and see if there is anything missing that'd be useful. The only scenario I could think of not covered is already timed out packets on ordered channels. This is tricky since by construction only 1 packet can be timed out at the moment (since the channel closes). Since ICS20 doesn't use ordered channels, I'd prefer to wait to fix until more than 1 packet can be timed out

**Comment by ancazamfir** [comment]
Opened #984 to check if the event has been previously handled by another relayer. This should reduce the number of retries to one.  I tested with two hermes instances and the `ft-transfer` command.

File: relayer/src/event/rpc.rs [hunk 1]
```diff
@@ -99,13 +111,16 @@ pub fn build_event(mut object: RawObject) -> Result<IbcEvent, BoxError> {
         // - "register" and "send" for ICS27
         // However the attributes are all prefixed with "send_packet" therefore the overwrite here
         // TODO: This need to be sorted out
-        "transfer" | "register" | "send" => {
+        "transfer"
+        | "register"
+        | "send"
```
Comment (colin-axner) [inline relayer/src/event/rpc.rs:116]: could we update this for ICS27 as well?

Comment (adizere) [reply relayer/src/event/rpc.rs]: Not sure if my understanding is correct. 09ad7b8237c16d38cfa76b6079291084cc0c039c Please let me know if otherwise!

**Comment by adizere** [comment]
Note to self: remember about this change to `zero_custom_fields` that may break chain upgrades: https://github.com/informalsystems/ibc-rs/pull/1223/files#r674113798

File: relayer/src/event/rpc.rs (hunk 1 above)
Comment (colin-axner) [reply relayer/src/event/rpc.rs]: Looks like [b912347](https://github.com/informalsystems/ibc-rs/pull/948/commits/b9123478151aed717b0e7f21d27fd2a592d7eaaf) resolved this

**Comment by adizere** [comment]
Some context on this PR: Even if we build Hermes with these new proto files (SDK 0.43 and ibc-go v1), the relayer should still be able to submit chain upgrade proposals and upgrade clients with 0.42-based chains (gaiad 5.0.5) for instance.

So merging this PR should not break any CLI and we should retain backwards compat with 0.42-based chains. We need to test for that.

Another thing is that we should update compatibility.rs to mark 0.43 as a compat. version.

**Comment by romac** [comment]
Running the following command using Hermes built from this branch and a v4.2.1 Gaia node yields:

```
❯ hermes -c ~/.hermes/gm-config.toml tx raw upgrade-chain network1 network2 07-tendermint-0 10000000 3000

Aug 12 14:53:15.911  INFO Message UpgradePlanOptions { src_chain_config: ChainConfig { id: ChainId { id: "network2", version: 0 }, rpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27030), path: "/", query: None, fragment: None }, scheme: Http, host: "localhost", port: 27030 }, websocket_addr: Url { inner: Url { scheme: "ws", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27030), path: "/websocket", query: None, fragment: None }, scheme: WebSocket, host: "localhost", port: 27030 }, grpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27032), path: "/", query: None, fragment: None }, scheme: Http, host: "localhost", port: 27032 }, rpc_timeout: 10s, account_prefix: "cosmos", key_name: "wallet", store_prefix: "ibc", max_gas: Some(500000), gas_adjustment: None, max_msg_num: MaxMsgNum(30), max_tx_size: MaxTxSize(2097152), clock_drift: 5s, trusting_period: 1209600s, trust_threshold: TrustThresholdFraction { numerator: 1, denominator: 3 }, gas_price: GasPrice { price: 0.001, denom: "stake" }, packet_filter: AllowAll }, dst_chain_config: ChainConfig { id: ChainId { id: "network1", version: 0 }, rpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27020), path: "/", query: None, fragment: None }, scheme: Http, host: "localhost", port: 27020 }, websocket_addr: Url { inner: Url { scheme: "ws", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27020), path: "/websocket", query: None, fragment: None }, scheme: WebSocket, host: "localhost", port: 27020 }, grpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(27022), path: "/", query: None, fragment: None }, scheme: Http, host: "localhost", port: 27022 }, rpc_timeout: 10s, account_prefix: "cosmos", key_name: "wallet", store_prefix: "ibc", max_gas: Some(500000), gas_adjustment: None, max_msg_num: MaxMsgNum(30), max_tx_size: MaxTxSize(2097152), clock_drift: 5s, trusting_period: 1209600s, trust_threshold: TrustThresholdFraction { numerator: 1, denominator: 3 }, gas_price: GasPrice { price: 0.001, denom: "stake" }, packet_filter: AllowAll }, src_client_id: ClientId("07-tendermint-0"), amount: 10000000, height_offset: 3000, upgraded_chain_id: ChainId { id: "network1", version: 0 }, upgraded_unbonding_period: None, upgrade_plan_name: "plan" }

Error:
   0: upgrade chain error
   1: tx response event consists of an error: deliver_tx on chain network1 for Tx hash 6FCF5B01B97296451F774BCD2DE8CEE47CA16E6F84CC3A6A5AE4F33249A1B05B reports error: code=Err(2), log=Log("unable to resolve type URL /ibc.core.client.v1.UpgradeProposal: tx parse error")
```

**Comment by hu55a1n1** [comment]
I have tested this branch against all gaia versions v4.0.3 onwards, i.e. SDK version >=0.41.3. All E2E tests are passing.

File: relayer/src/event/rpc.rs [hunk 2]
```diff
@@ -59,52 +59,68 @@ pub fn get_all_events(
...
-        "submit_misbehaviour" => Ok(IbcEvent::from(ClientEvents::ClientMisbehaviour::try_from(
-            object,
-        )?)),
+        "create_client" | ibc::ics02_client::msgs::create_client::TYPE_URL => Ok(IbcEvent::from(
+            ClientEvents::CreateClient::try_from(object)?,
+        )),
+        "update_client" | ibc::ics02_client::msgs::update_client::TYPE_URL => Ok(IbcEvent::from(
+            ClientEvents::UpdateClient::try_from(object)?,
+        )),
+        "submit_misbehaviour" | ibc::ics02_client::msgs::misbehavior::TYPE_URL => Ok(
```
Comment (adizere) [inline relayer/src/event/rpc.rs:68]: @ancazamfir I remember some TODO or a loose end with respect to misbehavior/evidence events or handling. Any thoughts?

**Review by adizere — COMMENTED** [review]
Left a few ideas for improvements, none are blockers.

File: modules/src/application/mod.rs [hunk 3]
```diff
@@ -1 +1,6 @@
 pub mod ics20_fungible_token_transfer;
+
+// TODO: These consts should move into the ICS27 namespace
+pub const ICS27_BANK_SEND_TYPE_URL: &str = "/cosmos.bank.v1beta1.MsgSend";
+pub const ICS27_SEND_TYPE_URL: &str = "/intertx.MsgSend";
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/application/mod.rs:5]: Later edit: these constants should be deleted with PR https://github.com/informalsystems/ibc-rs/pull/1172/files.

File: relayer/src/event/rpc.rs (hunk 2 above)
Comment (ancazamfir) [reply relayer/src/event/rpc.rs:68]: yes, this was wrt to testing misbehaviour which we don't cover in CI. Just tested now, all looks good!

**Review by romac — APPROVED** [review]

**Pull request merged**